worker.go

```diff
@@ -86,198 +79,160 @@ func newWorker(o workerOpt) (*worker, error) {
-			wait := backoff * 2
-			backingOffLock.RUnlock()
-			time.Sleep(wait)
-			upFunc.Do(func() {
-				backingOffLock.Lock()
-				defer backingOffLock.Unlock()
-				// if we come back to a stable state, reset the failure count
-				if backoff == minWorkerErrorBackoff {
-					failureCount = 0
-				}
-
-				// earn back the backoff over time
-				if failureCount > 0 {
-					backoff = max(backoff/2, 100*time.Millisecond)
-				}
-			})
-		}()
-
-		metrics.StartWorker(worker.fileName)
-
-		// Create main dummy request
-		r, err := http.NewRequest(http.MethodGet, filepath.Base(worker.fileName), nil)
-		if err != nil {
-			panic(err)
-		}
+func stopWorkers() {
+	close(workersDone)
+}
 
-		r, err = NewRequestWithContext(
-			r,
-			WithRequestDocumentRoot(filepath.Dir(worker.fileName), false),
-			WithRequestPreparedEnv(worker.env),
-		)
-		if err != nil {
-			panic(err)
-		}
+func drainWorkers() {
+	watcher.DrainWatcher()
+	stopWorkers()
+}
 
-		if c := logger.Check(zapcore.DebugLevel, "starting"); c != nil {
-			c.Write(zap.String("worker", worker.fileName), zap.Int("num", worker.num))
-		}
+func restartWorkers() {
+	workerRestartWG.Add(1)
+	defer workerRestartWG.Done()
+	for _, worker := range workers {
+		workerShutdownWG.Add(worker.num)
+	}
+	workersAreRestarting.Store(true)
+	stopWorkers()
+	workerShutdownWG.Wait()
+	workersDone = make(chan interface{})
+	workersAreRestarting.Store(false)
+}
 
-		if err := ServeHTTP(nil, r); err != nil {
-			panic(err)
-		}
+func getDirectoriesToWatch(workerOpts []workerOpt) []string {
+	directoriesToWatch := []string{}
+	for _, w := range workerOpts {
+		directoriesToWatch = append(directoriesToWatch, w.watch...)
+	}
+	return directoriesToWatch
+}
 
-		fc := r.Context().Value(contextKey).(*FrankenPHPContext)
+func (worker *worker) startNewThread() {
+	getInactivePHPThread().setActive(
+		// onStartup => right before the thread is ready
+		func(thread *phpThread) {
+			thread.worker = worker
+			thread.scriptName = worker.fileName
+			thread.requestChan = make(chan *http.Request)
+			thread.backoff = newExponentialBackoff()
+			worker.threadMutex.Lock()
+			worker.threads = append(worker.threads, thread)
+			worker.threadMutex.Unlock()
+			metrics.ReadyWorker(worker.fileName)
+		},
+		// beforeScriptExecution => set up the worker with a fake request
+		func(thread *phpThread) {
+			worker.beforeScript(thread)
+		},
+		// afterScriptExecution => tear down the worker
+		func(thread *phpThread, exitStatus int) {
+			worker.afterScript(thread, exitStatus)
+		},
+		// onShutdown => after the thread is done
+		func(thread *phpThread) {
+			thread.worker = nil
+			thread.backoff = nil
+		},
+	)
+}
 
-		// if we are done, exit the loop that restarts the worker script
-		if workersAreDone.Load() {
-			break
-		}
+func (worker *worker) beforeScript(thread *phpThread) {
+	// if we are restarting due to file watching, wait for all workers to finish first
+	if watcherIsEnabled && workersAreRestarting.Load() {
```

Comment:
Why does the watcher need to be enabled? Wouldn't we want this behavior in case of any restart?

Reply:
You are right, this check is redundant 👍